zapcore/sampler.go

```diff
@@ -81,17 +81,87 @@ func (c *counter) IncCheckReset(t time.Time, tick time.Duration) uint64 {
 	return 1
 }
 
+// SamplingDecision represents a decision made by sampler.
+type SamplingDecision uint8
+
+const (
+	// LogDropped means that a log was dropped.
+	LogDropped SamplingDecision = iota
+	// LogSampled means that a log was successfully sampled.
+	LogSampled
+)
+
+// optionFunc wraps a func so it satisfies the SamplerOption interface.
+type optionFunc func(*sampler)
+
+func (f optionFunc) apply(s *sampler) {
+	f(s)
+}
+
+// SamplerOption configures a Sampler option.
+type SamplerOption interface {
+	apply(*sampler)
+}
+
+// NopSamplingHook is the default hook used by sampler.
```

Comment:
Do we actually need to export this? We'll use this if unset.

Reply:
Maybe if someone wants to test this as NewSampler* constructors are public?

Reply:
NewSampler can be public but I mean the implementation of a no-op hook. Users don't really need access to that; they can write their own empty function, or just not provide the SamplingHook option.

Reply:
Agree, not even sure if we need this in the production code, we could keep it in the test file?

Reply:
It's production. The sampler defaults the hook field to NopSamplingHook.

Reply:
I see, in that case, can we unexport and keep it here?